### DISCUSSION TITLE
Can't seem to change button widths

### DISCUSSION THREAD

**Discussion by Jono1991** [TomSchimansky/CustomTkinter#490] [General]
https://pastebin.com/nzAG4YYT

So, I've got everything I want placed onto the window, but I can't seem to get 'width' to work on Line 51 and Line 58. I'm trying to shrink them so they look uniform and neat. Am I doing something wrong here?

**Comment by LineIndent** [comment]
I looked over the code you posted. First thing is maybe to consider using .pack() or .grid() managers for your layout, it just makes things a little more responsive and it makes placement of widgets easier. So I reworked the code you had and instead of .place() I used .pack( ):

```
import tkinter as tk
import customtkinter

customtkinter.set_appearance_mode("dark")
customtkinter.set_default_color_theme("blue")

HEADER_FONT = ("Calibri", 24)
SMALLER_FONT = ("Calibri", 12)


class App:
    def __init__(self):
        super().__init__()
        self.window = customtkinter.CTk()
        self.window.geometry("643x643")
        self.window.resizable(False, False)
        self.window.title("Prosperity Finances Banking Group")

        self.login_frame = customtkinter.CTkFrame(self.window, fg_color="grey25")
        self.login_frame.pack(side=tk.BOTTOM, fill=tk.BOTH, expand=False)

        self.login_header = customtkinter.CTkLabel(
            self.login_frame, text="Login", text_font=HEADER_FONT, fg_color="grey25"
        )
        self.login_header.pack(side=tk.TOP, fill=tk.X, expand=False, pady=10)

        # Main contianer
        self.container = customtkinter.CTkFrame(self.login_frame, fg_color="grey25")
        self.container.pack(side=tk.BOTTOM, fill=tk.BOTH, expand=True)

        # space
        self.space = customtkinter.CTkFrame(
            self.container, fg_color="grey25", height=10
        )
        self.space.pack()

        # Username widgets
        self.username_frame = customtkinter.CTkFrame(self.container, fg_color="grey25")
        self.username_frame.pack(pady=5, expand=True)

        self.login_username = customtkinter.CTkLabel(
            self.username_frame,
            text="Username: ",
            text_font=SMALLER_FONT,
            fg_color="grey25",
        )
        self.login_username.pack(side=tk.LEFT)

        self.login_entry = customtkinter.CTkEntry(self.username_frame, width=230)
        self.login_entry.pack(side=tk.LEFT, fill=tk.X)

        # space
        self.space = customtkinter.CTkFrame(
            self.container, fg_color="grey25", height=10
        )
        self.space.pack()

        # Password widgets
        self.passowrd_frame = customtkinter.CTkFrame(self.container, fg_color="grey25")
        self.passowrd_frame.pack(pady=5)

        self.login_password = customtkinter.CTkLabel(
            self.passowrd_frame,
            text="Password: ",
            text_font=SMALLER_FONT,
            fg_color="grey25",
        )
        self.login_password.pack(side=tk.LEFT)

        self.password_entry = customtkinter.CTkEntry(self.passowrd_frame, width=230)
        self.password_entry.pack(side=tk.LEFT, fill=tk.X)

        # space
        self.space = customtkinter.CTkFrame(
            self.container, fg_color="grey25", height=30
        )
        self.space.pack()

        # Button frame
        self.button_frame = customtkinter.CTkFrame(self.container)
        self.button_frame.pack()

        # space
        self.space = customtkinter.CTkFrame(
            self.button_frame, fg_color="grey25", width=60, height=10, bg_color="grey25"
        )
        self.space.pack(side=tk.LEFT, fill=tk.Y, expand=False)

        self.login_button = customtkinter.CTkButton(
            self.button_frame,
            text="Login",
            text_font=SMALLER_FONT,
            corner_radius=50,
            bg_color="grey25",
        )
        self.login_button.pack(side=tk.LEFT)

        # space
        self.space = customtkinter.CTkFrame(
            self.button_frame, fg_color="grey25", width=60, height=10, bg_color="grey25"
        )
        self.space.pack(side=tk.LEFT, fill=tk.Y, expand=False)

        # space
        self.space = customtkinter.CTkFrame(
            self.button_frame, fg_color="grey25", width=35, height=10, bg_color="grey25"
        )
        self.space.pack(side=tk.RIGHT, fill=tk.Y, expand=False)

        self.register_button = customtkinter.CTkButton(
            self.button_frame,
            text="Register",
            text_font=SMALLER_FONT,
            corner_radius=50,
            bg_color="grey25",
        )
        self.register_button.pack(side=tk.RIGHT)

        # space
        self.space = customtkinter.CTkFrame(
            self.container, fg_color="grey25", height=30
        )
        self.space.pack()

    def run(self):
        self.window.mainloop()


if __name__ == "__main__":
    app = App()
    app.run()


```
As for the button resizing option, the default is set to 140px I think and it doesn't seem to go below that number. Try placing a tkinter text inside a customtkinter button where ``` text=None ```

I found a video that does this with good UI and layout: https://www.youtube.com/watch?v=SbZ30zbXgmk

**Comment by Jono1991** [comment]
Hello,

Thanks for all of that information and rewriting my code to make more sense, that was really useful. I've been trying to figure out from Tom's documentation on how to create classes in custom modules in order to import them to your main app, but I just can't get it to work. It gives errors on importing all the other modules again, and then other errors I don't understand. I'll keep trying and tinkering about. But I made some other observations in the video you sent me and had some questions.

I noticed towards the end when using functions from his other modules like 'ui.color.MAIN', why does 'MAIN' appear in light blue and why is it all in capitals? Is it necessary to have those things?

And at the start, why does he create variables based on the size of the app, before even using calling .geometry()? I also didn't understand any of the code he used for that.

The frame he called self.container, that's not the same as a Toplevel is it? It's just a frame that has everything else in it.

**Reply by LineIndent** [reply]
Hey, 

Yeah at first glance the code can seem overwhelming, but I'll try and explain what's happening in the video. 

1. The ``` ui.Color.MAIN ```. What's happening here is a second file in the same folder, called ```ui.py``` is being imported that contains a class called ```Color``` in which inside that class there is a constant (all capital, i.e. cannot be changed) called ```MAIN``` and this constant is set to a color in hex form, such as ```"#232528"```. So the ```ui.py``` file will look something like this:
```
class Color:
    MAIN = "#f9e74d"
```
Therefore, you can use this color in the main python file or script by importing the file at the top like always ```import ui```, and then you can access the color and use it on the widgets as such ```fg_color = ui.Color.MAIN``` or any other widget paramter that accepts color. The name of the color can be anything you want. Also it doesn't have to be a constant (all capitals).

2. The sizing and geometry. What's happening here is that the window is being centered on the screen. To do that, we need to get the x and y coordinates (a little bit of math) and apply it to the ```.geometry()``` method. You can get a user's screen dimensions by calling ```self.winfo_width()``` or ```self.winfo_height()```. Getting the centered coordinates can be done the way the (X, Y) are calculated. 

3.The ```self.container```. This is just the name of the frame being created, not a Toplevel window. 

I hope this info helps!

**Comment by Jono1991** [comment]
Oh yeah I remember now, Programming with Mosh did a video about class attributes like that. So every class that inherits from that class will automatically have that constant (in this case MAIN, TITLE, etc). And your other answers clear up those things too, I don't think I particularly need to spawn the window to the centre of the screen so I'll look into doing that another time.

I also realised why I was failing to do the multiple windows thing, and I don't know if this is something that could be added to the documentation. I copied and pasted the example app code into `main.py` and the external frame into `newMain'py`. I didn't know why it wasn't working until it finally hit me, that I needed to prefix `RadioButtonFrame()` with the name of the external file. To make it work as is, I put this at the top - `from newMain import *`,

I don't suppose you could enlighten me as to how I can more information about the `header_name` keyword argument? I don't really get it or why it's there.

Edit: I'm struggling to get this working. I haven't updated to your rewrite yet, but I thought I figured out the problem, even though I haven't.

main.py - https://pastebin.com/HRxqSzvx
login.py - https://pastebin.com/rBesWseW

What am I doing wrong here?

Edit2:

So technically I got it working, but it just shows a part of a hand? I replaced `login.LoginScreen()` with `loginscreen = login.LoginScreen()` and put `loginscreen.pack()` under it. But it still looks weird?

**Reply by LineIndent** [reply]
There were a few mistakes in the code, but it works fine on my end.

_main.py_
```
import customtkinter
import login


class App(customtkinter.CTk):
    def __init__(self):
        super().__init__()

        self.geometry("643x643")
        self.resizable(False, False)
        self.title("Prosperity Finances Banking Group")

        login.LoginScreen()


if __name__ == "__main__":
    app = App()
    app.mainloop()

```

_login.py_

```
from tkinter import *
from PIL import ImageTk, Image
import customtkinter

customtkinter.set_appearance_mode("dark")
customtkinter.set_default_color_theme("green")

HEADER_FONT = ("Calibri", 24, "bold")
SMALLER_FONT = ("Calibri", 12, "bold")


class LoginScreen:
    def __init__(self):
        super().__init__()

        self.image_frame = customtkinter.CTkFrame(width=643, height=643)
        self.image_frame.place(anchor="nw", x=0, y=0)

        self.splash = ImageTk.PhotoImage(Image.open("splash_screen.jpg"))

        self.image_label = customtkinter.CTkLabel(self.image_frame, image=self.splash)
        self.image_label.place(anchor="center", x=643 // 2, y=643 // 2)

        self.login_frame = customtkinter.CTkFrame()
        self.login_frame.place(anchor="nw", height=214, width=643, x=0, y=429)

        self.login_header = customtkinter.CTkLabel(
            self.login_frame, text="Login", text_font=HEADER_FONT, padx=0, pady=0
        )
        self.login_header.place(anchor="center", x=322, y=29)

        self.login_username = customtkinter.CTkLabel(
            self.login_frame, anchor="w", text="Username: ", text_font=SMALLER_FONT
        )
        self.login_username.place(anchor="center", x=205, y=76)

        self.login_password = customtkinter.CTkLabel(
            self.login_frame, anchor="w", text="Password: ", text_font=SMALLER_FONT
        )
        self.login_password.place(anchor="center", x=204, y=124)

        self.login_entry = customtkinter.CTkEntry(self.login_frame, width=230)
        self.login_entry.place(anchor="center", x=373, y=75)

        self.password_entry = customtkinter.CTkEntry(
            self.login_frame, width=230, show="*"
        )
        self.password_entry.place(anchor="center", x=373, y=125)

        self.login_button = customtkinter.CTkButton(
            self.login_frame,
            text="Login",
            text_font=SMALLER_FONT,
            corner_radius=30,
            width=120,
        )
        self.login_button.place(anchor="center", x=249, y=175)

        self.register_button = customtkinter.CTkButton(
            self.login_frame,
            text="Register",
            text_font=SMALLER_FONT,
            corner_radius=30,
            width=120,
        )
        self.register_button.place(anchor="center", x=395, y=175)
```

The issue was that 1) you were calling ```customtkinter.CTkFrame``` in the ```login.py``` class, which is wrong for two reasons: A) You're not just using frames in that class, you're using labels, etc, and B) you don't need to call anything in that class because you already call the entire ```customtkinter.CTk``` library in the ```main.py```. The second issue was that in ```login.py```, some of the widgets had self as there location, and this generated an extra window. So I removed the ```self``` keyword and when I ran the code from ```main.py```, it worked.

**Comment by Jono1991** [comment]
I thought all frames had to inherit from `customtkinter.CTkFrame`, I also didn't know you could create a frame without specifying it's location like you rewrote on `self.image_frame`. Lots of new concepts I'm learning here.

So, what should I make note of here? Because I'm still not 100% sure how this works. So if I call `customtkinter.CTk` in the main app, then I don't need to inherit anything from a frame in a different module? How come inheriting `customtkinter.CTkFrame` invalidates labels, buttons, etc?

**Reply by LineIndent** [reply]
We don't have to specify `self` because we are creating an instance of the `class login`, so we would only specify it's location when we call it. If you're calling `CTkFrame` in the `class login`, I would assume that it's the only thing you can use in the GUI build, and not the entire library. It doesn't necessary invalidate so much as it simply doesn't access them. It's similar to importing a module like `from PIL import ImageTk, Image`, in which only two methods of the entire Pillow library are imported instead of everything, which would be denoted as `from PIL import *`.
